src/EventStore.Core/TransactionLog/Scavenging/ParallelLoop.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EventStore.Common.Utils;
+
+namespace EventStore.Core.TransactionLog.Scavenging {
+	public static class ParallelLoop {
+		// passes each item in `source` to `process`, according to the `degreeOfParallelism.
+		// processing is never done on the calling thread
+		// emitCheckpoint is called on the calling thread with the latest checkpoint that is complete
+		// items are queried for checkpoint
+		//   - getCheckpointInclusive returns the checkpoint that can be emitted when all items up to
+		//     and including this one have been completed.
+		//   - getCheckpointExclusive returns the checkpoint that can be emitted when all items up to
+		//     but excluding this one have been completed.
+		public static void RunWithTrailingCheckpoint<T>(
+			IEnumerable<T> source,
+			int degreeOfParallelism,
+			Func<T, int> getCheckpointInclusive,
+			Func<T, int?> getCheckpointExclusive,
+			Action<int, T> process,
+			Action<int> emitCheckpoint,
+			Action onConsiderEmit = null) {
+
+			Ensure.Positive(degreeOfParallelism, nameof(degreeOfParallelism));
+
+			// in each slot the store the checkpoint that can be emitted when every item before the one
```

Comment:
typo in `each slot the store the checkpoint`

Reply:
fixed